### DISCUSSION TITLE
Cannot Save Files to Browser

### DISCUSSION THREAD

**Discussion by kalshair** [sebastienros/jint#1299] [Q&A] [answered]
Hello, 
I am trying to write a file to the browser using SheetJS's standalone scripts, but Jint is throwing an exception that it cannot save the file. Is there anything specific I have to do to allow Jint to write files to the client browser?
Exception is as follows: 
`Exception thrown: 'Jint.Runtime.JavaScriptException' in Jint.dll`
`An exception of type 'Jint.Runtime.JavaScriptException' occurred in Jint.dll but was not handled in user code`
`cannot save file [filename].xlsx`

Unfortunately, I am unable to share the code itself as it is all considered closed-source, proprietary, and company property. However, I can describe what it is I'm doing: 
Currently, I am reading all of my scripts as if they were text files into strings, and then executing them within a new Engine. I am then invoking the Javascript function in which I am trying to write files with.

The Javascript simply takes an array of arrays, creates an Excel sheet out of them, and then tries to write the workbook to a new Excel file (.xlsx, specifically).

The scripts I am consuming (in order) are the full sheetjs browser script, and then my full script. It seems all of the functions on the Javascript end are working, apart from writing to a file.

**Comment by sebastienros** [comment]
> apart from writing to a file

Jint doesn't know how to write files, on purpose. You need to teach it how to do it. For instance by implement the code your script is calling do create these ... and do it in a safe way such that malicious scripts can't harm your system (isolated folder with limits).

**Comment by kalshair** [comment]
Could you elaborate more on what you mean by 'teach it'?
SheetJS tries to save to the browser (and thus download the file to the client computer). Do I have to rework my Javascript code? Or do I have to rework my .NET code?

**Comment by sebastienros** [comment, marked as answer]
> SheetJS tries to save to the browser

SheetJS calls into the browser API. This is the part you need to add to the Engine, in .NET. So if SheetJS calls an API named `file.save()` you need to create a property `file` that has a `save()` method that will call into your .NET code.

**Reply by kalshair** [reply]
I really appreciate the answer. Do you think you could provide a short example of how that would look in code?
